### PR TITLE
[neighorch]: Replace VOQ remote nexthop instead of skipping addNeighbor

### LINKED ISSUE #27585: Bug: [Multi ASIC] PortChannel fails to delete on remote ASIC after quick IP/Interface reassignment to another PortChannel
### Is it platform specific

generic

### Importance or Severity

Critical

### Description of the bug

Deleting a PortChannel (Po1) and reassigning its IP and interface to another PortChannel (Po2) before deletion clean-up has completed causes Po1 to fail to delete on the remote ASIC and leads to traffic drops

Issue revealed by traffic drops during test_po_update_io_no_loss on multi asic system, caused by test_po_update running before it and leaving the system in a bad state.

#### We can reproduce with test_po_update.py on a multi asic box
1. test_po_update.py::test_po_update runs, creates PortChannel999 using the interfaces and address of an existing PortChannel. This is already enough, the problem has been encountered at this point.
3. test_po_update.py::test_po_update_io_no_loss runs, if it (randomly) picks a src and dest portchannel on 2 different asics then the test will fail since traffic will drop due to the discrepancy of PortChannel 999 on each asic.

#### Or Manually by using steps similar to test_po_update to replicate the issue:

1) Remove eth8 and IP from existing Po2
2) wait for BGP, 30-40s
3) create Po1
4) give Po1 the interface and IP from Po2
5) wait for BGP
6) remove ip, interface from Po1
7) delete Po1
8) re-add eth and ip to Po2

Now we will have a discrepancy between asic0 and asic1, Po1 fails to delete on asic1.

Example:
```
2026-05-28 18:57:03 | + PC=PortChannel101
2026-05-28 18:57:03 | ++ show ip interfaces
2026-05-28 18:57:03 | ++ awk -v pc=PortChannel101 '$1==pc {print $2; exit}'
2026-05-28 18:57:06 | + PC_IP=10.0.0.0/31
2026-05-28 18:57:06 | + sudo config portchannel -n asic0 member del PortChannel101 Ethernet8
2026-05-28 18:57:06 | + sudo config interface -n asic0 ip remove PortChannel101 10.0.0.0/31
2026-05-28 18:57:07 | + sleep 40

2026-05-28 18:57:47 | + sudo config portchannel -n asic0 add PortChannel922
2026-05-28 18:57:47 | + sudo config portchannel -n asic0 member add PortChannel922 Ethernet8
2026-05-28 18:57:48 | + sudo config interface -n asic0 ip add PortChannel922 10.0.0.0/31
2026-05-28 18:57:48 | + sleep 40

2026-05-28 18:58:28 | + sudo config interface -n asic0 ip remove PortChannel922 10.0.0.0/31
2026-05-28 18:58:29 | + sudo config portchannel -n asic0 member del PortChannel922 Ethernet8
2026-05-28 18:58:30 | + sudo config portchannel -n asic0 del PortChannel922

2026-05-28 18:58:31 | + sudo config interface -n asic0 ip add PortChannel101 10.0.0.0/31
2026-05-28 18:58:33 | + sudo config portchannel -n asic0 member add PortChannel101 Ethernet8
2026-05-28 18:58:34 | + sleep 3
```

We will see see these in syslog
```
2026 May 28 18:59:05.752834 NOTICE swss0#orchagent: :- removeLag: Remove LAG PortChannel922 lid:200
```
Repeatedly:
```
2026 May 28 18:59:05.753281 ERR swss1#orchagent: :- removeLag: Failed to remove ref count 1 LAG HOST|asic0|PortChannel922
```

### Actual Behavior and Expected Behavior

swss1 fails to delete Po2 because the neighbours are never cleaned up and it always has ref count 1/
We expect asic0 and asic1 to synchronize, the PortChannel should get deleted from both asics.

### Relevant log output

```shell
# syslogs taken from a run of test_po_update.py

# test po update starts

# 1. remove Ethernet8 from PortChannel101
2026 May 27 18:53:51.050642 HOSTNAME INFO python3.13[304839]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config portchannel -n asic0 member del PortChannel101 Ethernet8 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# orchagents react
2026 May 27 18:53:51.617497 HOSTNAME NOTICE swss0#orchagent: :- removeLagMember: Remove member Ethernet8 from LAG PortChannel101 lid:20000000007a4 lmid:1b0000000007aa

2026 May 27 18:53:51.629751 HOSTNAME NOTICE swss1#orchagent: :- removeLagMember: Remove member HOSTNAME|asic0|Ethernet8 from LAG HOSTNAME|asic0|PortChannel101 lid:1020000000007ae lmid:11b00000000088c


# 2. config new PortChannel999 for asic0
2026 May 27 18:54:28.064607 HOSTNAME INFO python3.13[306636]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config portchannel -n asic0 add PortChannel999 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# swss0 orchagent reacts
2026 May 27 18:54:28.425571 HOSTNAME NOTICE swss0#orchagent: :- addLag: Create an empty LAG PortChannel999 lid:2000000000a53

# swss1 orchagent reacts
2026 May 27 18:54:28.426897 HOSTNAME NOTICE swss1#orchagent: :- addLag: Create an empty LAG HOSTNAME|asic0|PortChannel999 lid:102000000000a6e

# 3. add Ethernet8 to PortChannel999
2026 May 27 18:54:28.811224 HOSTNAME INFO python3.13[306669]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config portchannel -n asic0 member add PortChannel999 Ethernet8 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# orchagents react
2026 May 27 18:54:29.495831 HOSTNAME NOTICE swss0#orchagent: :- addLagMember: Add member Ethernet8 to LAG PortChannel999 lid:2000000000a53 pid:1000000000002

2026 May 27 18:54:29.497336 HOSTNAME NOTICE swss1#orchagent: :- addLagMember: Add member HOSTNAME|asic0|Ethernet8 to LAG HOSTNAME|asic0|PortChannel999 lid:102000000000a6e pid:15d000000000044

# 4. add an ip address to the PC
2026 May 27 18:54:29.648000 HOSTNAME INFO python3.13[306711]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config interface -n asic0 ip add PortChannel999 10.0.0.0/31 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None


# orchagents react
2026 May 27 18:54:30.021689 HOSTNAME NOTICE swss0#orchagent: :- addRouterIntfs: Create router interface PortChannel999 MTU 9100
2026 May 27 18:54:30.023910 HOSTNAME NOTICE swss1#orchagent: :- addRouterIntfs: Create router interface HOSTNAME|asic0|PortChannel999 MTU 1492

# orch adds neighbours, nexthop, nexthopgroup
2026 May 27 18:54:31.507949 HOSTNAME NOTICE swss0#orchagent: :- addNeighbor: Created neighbor ip 10.0.0.1, 0e:41:01:83:39:33 on PortChannel999
2026 May 27 18:54:31.508870 HOSTNAME NOTICE swss0#orchagent: :- addNextHop: Created next hop 10.0.0.1 on PortChannel999
2026 May 27 18:54:31.510671 HOSTNAME NOTICE swss1#orchagent: :- addNeighbor: Created neighbor ip 10.0.0.1, 0e:41:01:83:39:33 on HOSTNAME|asic0|PortChannel999
2026 May 27 18:54:31.511571 HOSTNAME NOTICE swss1#orchagent: :- addNextHop: Created next hop 10.0.0.1 on Ethernet-IB1

2026 May 27 18:54:32.092776 HOSTNAME NOTICE swss0#orchagent: :- addNextHopGroup: Create next hop group 10.0.0.1@PortChannel999,10.0.0.3@PortChannel103,10.0.0.5@PortChannel105,10.0.0.7@PortChannel110,10.0.0.9@PortChannel112,10.0.0.11@PortChannel114,10.0.0.13@Ethernet-IB0,10.0.0.15@Ethernet-IB0,10.0.0.21@Ethernet-IB0,10.0.0.29@Ethernet-IB0
2026 May 27 18:54:32.189960 HOSTNAME NOTICE swss0#orchagent: :- addNextHopGroup: Create next hop group 10.0.0.1@PortChannel999,10.0.0.3@PortChannel103,10.0.0.5@PortChannel105,10.0.0.7@PortChannel110

2026 May 27 18:54:32.538159 HOSTNAME NOTICE swss1#orchagent: :- addNextHopGroup: Create next hop group 10.0.0.1@Ethernet-IB1,10.0.0.3@Ethernet-IB1,10.0.0.5@Ethernet-IB1,10.0.0.7@Ethernet-IB1,10.0.0.9@Ethernet-IB1,10.0.0.11@Ethernet-IB1,10.0.0.13@PortChannel116,10.0.0.15@PortChannel123,10.0.0.21@PortChannel118,10.0.0.29@PortChannel125
2026 May 27 18:54:34.070886 HOSTNAME NOTICE swss1#orchagent: :- addNextHopGroup: Create next hop group 10.0.0.1@Ethernet-IB1,10.0.0.3@Ethernet-IB1,10.0.0.5@Ethernet-IB1,10.0.0.7@Ethernet-IB1


# test po update cleanup start

# 1. remove the portchannel999's ip
2026 May 27 18:55:06.634946 HOSTNAME INFO python3.13[307926]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config interface -n asic0 ip remove PortChannel999 10.0.0.0/31 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# swss0 orchagent reacts, this fail to remove repeats a long time
2026 May 27 18:55:07.394728 HOSTNAME NOTICE swss0#orchagent: :- removeRouterIntfs: Router interface PortChannel999 is still referenced with ref count 3

# 2. remove interface Ethernet8 from PortChannel999
2026 May 27 18:55:12.750586 HOSTNAME INFO python3.13[308053]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config portchannel -n asic0 member del PortChannel999 Ethernet8 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# both orchs react
2026 May 27 18:55:13.711219 HOSTNAME NOTICE swss0#orchagent: :- removeLagMember: Remove member Ethernet8 from LAG PortChannel999 lid:2000000000a53 lmid:1b000000000a54
2026 May 27 18:55:14.170982 HOSTNAME NOTICE swss1#orchagent: :- removeLagMember: Remove member HOSTNAME|asic0|Ethernet8 from LAG HOSTNAME|asic0|PortChannel999 lid:102000000000a6e lmid:11b000000000a6f


# 3. add ip to portchannel101
2026 May 27 18:55:20.991520 HOSTNAME INFO python3.13[308858]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config interface -n asic0 ip add PortChannel101 10.0.0.0/31 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# 4. add ethernet8 to pc101
2026 May 27 18:55:21.804078 HOSTNAME INFO python3.13[308950]: ansible-ansible.legacy.command Invoked with _raw_params=sudo config portchannel -n asic0 member add PortChannel101 Ethernet8 _uses_shell=True expand_argument_vars=True stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None

# both orchs react
2026 May 27 18:55:23.451943 HOSTNAME NOTICE swss0#orchagent: :- addLagMember: Add member Ethernet8 to LAG PortChannel101 lid:20000000007a4 pid:1000000000002
2026 May 27 18:55:23.454114 HOSTNAME NOTICE swss1#orchagent: :- addLagMember: Add member HOSTNAME|asic0|Ethernet8 to LAG HOSTNAME|asic0|PortChannel101 lid:1020000000007ae pid:15d000000000044




# there's no addRouterIntfs call but neighbours added
# supposedly there is no addRouterIntfs call because setIntf doesn't need to make a new RIF since there is one still alive from the Ipv6 prefix we didn't remove.
2026 May 27 18:55:24.924527 HOSTNAME NOTICE swss0#orchagent: :- addNeighbor: Created neighbor ip 10.0.0.1, 0e:41:01:83:39:33 on PortChannel101
2026 May 27 18:55:24.925495 HOSTNAME NOTICE swss0#orchagent: :- addNextHop: Created next hop 10.0.0.1 on PortChannel101


# 5. the test reaches its END
2026 May 27 18:55:38.188476 HOSTNAME NOTICE root: PBAILEY END test_po_update MARKER asic0



# Finally the last 2 neighbours removed and step #1 completes
2026 May 27 18:55:53.813472 HOSTNAME NOTICE swss0#orchagent: :- removeRouterIntfs: Router interface PortChannel999 is still referenced with ref count 2
2026 May 27 18:55:53.814071 HOSTNAME NOTICE swss0#orchagent: :- removeNeighbor: Removed next hop 10.0.0.1 on PortChannel999
2026 May 27 18:55:53.815110 HOSTNAME NOTICE swss0#orchagent: :- removeNeighbor: Removed neighbor 0e:41:01:83:39:33 on PortChannel999
2026 May 27 18:55:53.815651 HOSTNAME ERR swss0#orchagent: :- removeLag: Failed to remove ref count 1 LAG PortChannel999
2026 May 27 18:55:53.820779 HOSTNAME NOTICE swss0#orchagent: :- removeRouterIntfs: Remove router interface for port PortChannel999
2026 May 27 18:55:53.823867 HOSTNAME NOTICE swss0#orchagent: :- removeLag: Remove LAG PortChannel999 lid:2000000000a53

# which notifies swss1, and here we have the issue
2026 May 27 18:55:54.015969 HOSTNAME ERR swss1#orchagent: :- removeLag: Failed to remove ref count 1 LAG HOSTNAME|asic0|PortChannel999
# this repeats all the way through the log file even during the next test. we NEVER see "swss1#orchagent: :- removeNeighbor" for PortChannel999

2026 May 27 18:56:28.737185 HOSTNAME ERR swss1#orchagent: :- removeLag: Failed to remove ref count 1 LAG HOSTNAME|asic0|PortChannel999
```

### Output of `show version`, `show techsupport`

```shell
202511
```

### Attach files (if any)

_No response_

### PR DESCRIPTION
**Pull request by peterbailey-arista**
Fixes https://github.com/sonic-net/sonic-buildimage/issues/27585 (issue)

**What I did**
This fix removes the skip from [PR3329](https://github.com/sonic-net/sonic-swss/pull/3329
) and lets addNeighbor handle the case where a nexthop already exists for a remote port.
It removes the old SAI nexthop (or tracks it as stale for later cleanup if still in use), then creates a new one which preserves the existing ref_count and nh_flags.

**Why I did it**

The issue seen was that when we remove the ip address from lag A and add it to lag B, then attempt to delete lag A, we can not do so because the routes are added and converge for lag B before lag A is removed. On the remote asic the system ports both use Ethernet-IBX as the alias for the nexthops, meaning that lag A fails to delete when it sees that there are still references to it.

removeLag fails because removeRouterIntfs has not completed. 
removeRouterIntfs fails because removeNeighbor has not completed. 
removeNeighbor fails because nexthop "10.0.0.1 on Ethernet-IB1" exists

This nexthop was never deleted because [PR3329](https://github.com/sonic-net/sonic-swss/pull/3329
) added a skip that prevents doVoqSystemNeighTask from calling addNeighbor if there is already a nexthop associated to (10.0.0.1, Ethernet-IB1), which there is from lag B.

In the past, before [PR3329](https://github.com/sonic-net/sonic-swss/pull/3329
), when addNeighbor was called it would simply clobber the current nexthop with a new one. This effectively reset the ref count to 0 and allowing the lag to be deleted

m_syncdNextHops[nexthop] = next_hop_entry

[PR3329](https://github.com/sonic-net/sonic-swss/pull/3329
) results in 2 issues,
1. removeLag becomes impossible on the remote asic for a remote port that uses the same ip as another remote port.
2. the stale neighbor and nexthop remains for lag A and is never created for lag B.

**How I verified it**

I reran the pc/ sonic-mgmt test suite on my multi-asic fixed testbeds to see they no longer failed. Ran the entire sonic-mgmt suite on these boxes as well without regression.

I also used custom scripting that could replicate the issue consistently to verify that the lag id discrepancy between asics was solved by my fix.

I added a `VoqRemoteNeighborReplaceAndRemove` unit test case as well.

**Details if related**

### CODE REVIEW

File: orchagent/neighorch.cpp [hunk 1]
```diff
@@ -1643,7 +1693,49 @@ bool NeighOrch::removeNeighbor(NeighborContext& ctx, bool disable)
     SWSS_LOG_INFO("Try to remove neighbor %s on %s",
                    ip_address.to_string().c_str(), alias.c_str());
 
-    if (m_syncdNextHops.find(nexthop) != m_syncdNextHops.end() && m_syncdNextHops[nexthop].ref_count > 0)
+    /*
+        on VOQ the nexthop keyed by (ip, inband port) can exist for a multiple
```
[the remark this answers is not in the document]
Comment (peterbailey-arista) [reply orchagent/neighorch.cpp]: 'a' removed. Done

File: orchagent/neighorch.cpp [hunk 2]
```diff
@@ -378,7 +378,48 @@ bool NeighOrch::addNextHop(NeighborContext& ctx)
...
+                        m_intfsOrch->decreaseRouterIntfsRefCount(entry.alias);
+                        decCrmNextHopCounter(entry.ip_address);
+                    }
+                    else
+                    {
+                        // track the sai nexthop for deletion later
+                        SWSS_LOG_NOTICE("Unable to clear nexthop %s for %s to create for %s, sai rv:%d",
+                            nexthop.to_string().c_str(), entry.alias.c_str(), nh.alias.c_str(), status);
+                        m_staleNextHops[entry] = m_syncdNextHops[nexthop];
+                    }
```
[the remark this answers is not in the document]
Comment (peterbailey-arista) [reply orchagent/neighorch.cpp:416]: Accepting this but I have instead opted to reuse the codepath for SAI_STATUS_SUCCESS

File: orchagent/neighorch.cpp [hunk 3]
```diff
@@ -1643,7 +1693,49 @@ bool NeighOrch::removeNeighbor(NeighborContext& ctx, bool disable)
...
+                        m_intfsOrch->decreaseRouterIntfsRefCount(alias);
+                        decCrmNextHopCounter(neighborEntry.ip_address);
+                    }
+                    // if we can't remove it then we need to wait to remove *this* neighbor
+                    else
+                    {
+                        SWSS_LOG_NOTICE("Unable to remove stale nexthop %s on %s during removal of %s, sai rv:%d",
+                            nexthop.to_string().c_str(), otherEntry.alias.c_str(), alias.c_str(), status);
+                        return false;
+                    }
```
[the remark this answers is not in the document]
Comment (peterbailey-arista) [reply orchagent/neighorch.cpp:1742]: Same here, I am accepting this but I have instead opted to reuse the codepath for SAI_STATUS_SUCCESS instead of the suggested code snippet.

**Comment by peterbailey-arista** [comment]
@deepak-singhal0408  I have gone ahead and incorporated copilot's suggestions - though the specific edge case it has outlined would be very hard for me to test. It implies to me that the nexthop was deleted from outside the orchagent. 
When the NH fails to delete it is usually with a SAI_STATUS_OBJECT_IN_USE hence why I am tracking it for deletion later

File: orchagent/neighorch.cpp [hunk 4]
```diff
@@ -378,7 +378,54 @@ bool NeighOrch::addNextHop(NeighborContext& ctx)
...
+            {
+                NeighborEntry entry = neigh.first;
+                if (entry.ip_address == nh.ip_address && entry.alias != nh.alias
+                        && m_intfsOrch->isRemoteSystemPortIntf(entry.alias))
+                {
+                    // 2. attempt to delete it from sai
+                    starting_ref_count = m_syncdNextHops[nexthop].ref_count;
+                    starting_nh_flags = m_syncdNextHops[nexthop].nh_flags;
+
+                    sai_status_t status = sai_next_hop_api->remove_next_hop(m_syncdNextHops[nexthop].next_hop_id);
```
Comment (saksarav-nokia) [inline orchagent/neighorch.cpp:399]: You are removing the nexthop here. How about if any routes were added with this nexthop?

File: tests/mock_tests/neighorch_ut.cpp [hunk 5]
```diff
@@ -474,4 +528,76 @@ namespace neighorch_test
         EXPECT_FALSE(gNeighOrch->isHwConfigured(VLAN1000_NEIGH));
         EXPECT_TRUE(gNeighOrch->isHwConfigured(VLAN2000_NEIGH));
     }
+
+    TEST_F(NeighOrchTest, VoqRemoteNeighborReplaceAndRemove)
```
Comment (saksarav-nokia) [inline tests/mock_tests/neighorch_ut.cpp:532]: Don't you need any new/update tests in test_virtual_chassis.py to test in Voq environment?

File: orchagent/neighorch.cpp (hunk 4 above)
Comment (peterbailey-arista) [reply orchagent/neighorch.cpp:399]: Thanks for reviewing @saksarav-nokia !

I just want to clarify your concern here: if routes were added with this nexthop then it won't be eligible for delete from SAI yet (`SAI_STATUS_OBJECT_IN_USE` or similar) - in which case the removal will fail and we will enter the else block. We then track `m_staleNextHops` until the routes have been cleaned up and allow for the nexthop to be cleaned up properly. 

This is the situation we're trying to avoid with this PR. Prior to the change we would retain a stale nexthop and neighbor and this would prevent the router interface from being deleted, preventing the remote lag/remote interface from being removed.

Is there a specific situation or edge case you have in mind? Or does this address your concern? 
Thanks again

File: tests/mock_tests/neighorch_ut.cpp (hunk 5 above)
Comment (peterbailey-arista) [reply tests/mock_tests/neighorch_ut.cpp:532]: I would need to look into these tests more if you think we need additional coverage beyond the mgmt tests that exposed the issue plus this new unit test

Comment (kenneth-arista) [reply tests/mock_tests/neighorch_ut.cpp:532]: I think any additional VS based test can come as a separate PR from this PR. This PR has included a mock test to provide the code coverage

**Review by deepak-singhal0408 — COMMENTED** [review]
@peterbailey-arista, Thanks for tackling this — the replace approach does fix the shipped single-move case. I want to flag a few design-level concerns, because I think the current mechanism is correct only under assumptions that don't always hold. Sharing some examples below.

The root of my concern: nexthop bookkeeping collapses both remote owners of an IP onto one key `(ip, inband)` and then decides which system-port owns it **implicitly, by event order** (the last `addNextHop` wins; the displaced OID goes into `m_staleNextHops`). That works when events arrive in the expected order and someone else reprograms the routes — but neither is guaranteed.

**1. Teardown in the "wrong" order → blackhole + stuck neighbor + leak.**
IP `10.0.0.1` has moved A→B, so `m_syncdNextHops[(10.0.0.1, IB0)] = B_oid` and `m_staleNextHops` holds `A_oid`. Now the **active owner B is removed before A** (move reverted / lag-B flap / out-of-order `SYSTEM_NEIGH` withdrawals):
- `removeNeighbor(B)` sees A still sharing the IP → `nexthop_still_valid = true` → skips the nexthop teardown, but `m_staleNextHops` holds A, not B, so nothing is cleaned; the map still points at `B_oid` while B's SAI neighbor is deleted → **active nexthop resolves through a dead system-port B (blackhole)**. A is never promoted back to active.
- Later `removeNeighbor(A)`: no other sharer → `nexthop_still_valid = false` → `ref_count > 0` → returns false → **A is undeletable** (the original symptom returns). When routes finally drain, `removeNextHop` frees `B_oid`, not `A_oid` → **`A_oid` + `rif(A)` leaked, `m_staleNextHops[A]` orphaned.**

**2. Three+ transient owners (N>2).**
Both the replace loop in `addNextHop` and the still-valid loop in `removeNeighbor` take the *first* matching other neighbor and `break`. A single slot + single `m_staleNextHops` entry can't represent 3 owners (e.g. A→B→C while reconvergence lags), so the rif/CRM decrement can be applied to a different neighbor than the OID actually removed → **CRM miscount + rif skew + a dropped OID.**

**3. Routes are never reprogrammed on the move.**
After A→B the map holds `B_oid`, but already-programmed routes still bind `A_oid` — routeorch caches the OID at bind time and doesn't re-read it. So correctness of even the happy path depends on **FRR re-advertising** the routes to rebind them, and until they drain, `remove_next_hop(A_oid)` returns `OBJECT_IN_USE`. If reconvergence is delayed/incomplete, A stays stuck. This also means the fix leans on external churn rather than doing the rebind itself.

**Suggested direction.**
The issue is that ownership is implicit and OID mutation is silent. A more robust shape, without collapsing the key:
- keep nexthop entries keyed by the real `(ip, system_port)` (A and B stay distinct — no clobber, no stale side-map),
- add one explicit resolver `(ip, inband) → active (ip, system_port)` that route lookup goes through, and
- on any owner change, **propagate it via the existing nexthop-invalidation path** (the same `NHFLAGS_IFDOWN` / `invalidNextHopInNextHopGroup()` fan-out) so dependent routes/NHG members are rebound — instead of relying on FRR churn.

That makes the behavior order-independent (removing the active owner just promotes the survivor), N-safe, and self-contained on rebind. Let's have a sync up on this.

**Comment by peterbailey-arista** [comment]
Thank you for reviewing! @deepak-singhal0408 I will take some time to understand your concerns and proposal
